test/RSocketClientTest.cpp

```diff
@@ -0,0 +1,33 @@
+// Copyright 2004-present Facebook. All Rights Reserved.
+
+#include "RSocketTests.h"
+
+#include <folly/io/async/ScopedEventBaseThread.h>
+#include <gtest/gtest.h>
+
+#include "rsocket/transports/tcp/TcpConnectionFactory.h"
+
+using namespace rsocket;
+using namespace rsocket::tests;
+using namespace rsocket::tests::client_server;
+
+TEST(RSocketClient, ConnectFails) {
+  folly::ScopedEventBaseThread worker;
+
+  folly::SocketAddress address;
+  address.setFromHostPort("localhost", 1);
+  auto client = RSocket::createConnectedClient(
+      std::make_unique<TcpConnectionFactory>(*worker.getEventBase(),
+                                             std::move(address)));
+
+  folly::Baton<> baton;
+  client.then([&](auto&) {
+    FAIL() << "the test needs to fail";
+    baton.post();
+  }).onError([&](const std::exception&) {
+    LOG(INFO) << "connection failed as expected";
+    baton.post();
+  });
```

Comment:
Can't you do `.get()` on the future instead of using a manual `folly::Baton`?  Or was that hitting the ASAN issues?  (I've hit them before too and they're weird...)